### ISSUE TITLE
Lubuntu + Calamares mount tmpfs on /tmp

### ISSUE THREAD

**Issue by hadmut** [calamares/calamares#1818]
Hi, 

I've reported this bug to Ubuntu/Lubuntu some time ago, https://bugs.launchpad.net/ubuntu/+source/calamares/+bug/1888526 but they see it rather as a calamares specific bug. Still exists in Lubuntu 21.10. 

**Describe the bug**

After installing Lubuntu Desktop (which uses Calamares), a tmpfs is mounted on /tmp. 

This is not just a violation of Linux standards, it also breaks lots of functionalities, since it limits the space on /tmp, consumes RAM, and other problems. This really breaks many functions of a running Linux system.  

**To Reproduce**

Install Lubuntu and do a df. 

**Expected behavior**

Just nothing mounted on /tmp, unless manually chosen by admin during installation time.

**Comment by kkofler** [comment]
As I understand it, this is a distro decision.

**Comment by adriaandegroot** [comment]
There is specific code to do that; if your / is on an SSD, /tmp is a tmpfs. It's been like that since 2014. See *fstab* module, `main.py` lines 204ff. This says nothing about good-or-bad-idea, just that "it's always been like that". 

I'm a bit disappointed that you (@hadmut) have to report this upstream, rather than the Lubuntu people themselves. They hang around in some of the chat channels for Calamares. None of the discussion in Launchpad is all that meaningful: it's just that you end up with tmpfs on `/tmp` rather than something else. Note, too, the competing Launchpad issue for making `/tmp` tmpfs -- from 2016.

I'm not a big fan of making **every** knob twiddle-able inside Calamares. For users with special uses for `/tmp`, working-around the installation choice is a single '#' added to `/etc/fstab'.

**Comment by hadmut** [comment]
I never knew that this is based on the detection of an SSD. I'd rather prefer to set the noatime flag for SSDs. 

Nevertheless, I just noticed that the recent version (Lubuntu 21.10) creates a swap file, although very small, where, to the best of my knowledge, tmpfs is able to move things. In former versions (Lubuntu 20.04) it created a tmpfs on /tmp while not having a swap file, which every now and then crashed systems, because both /tmp and RAM went full and system ran out of available mem. It can really cause severe problems. 

I'd really prefer to have some interaction telling „hey, we've detected an SSD, and we recommend
- tmpfs on /tmp
- noatime on filesystems
- tmpfs only with sufficiently large swap space.

or simple always this dialog (with a SSD detected - recommended comment). 

Some time ago, when I first found that tmpfs on /tmp, it was because I was post-mortem-analyzing, why a system had crashed. 

And if it has always been there since 2014: Today's ssd are more robust to multiple writes than 2014's ssds. 

I'd rather see problems with microsd cards and /dev/mmcblk flash chips e.g. in raspberrys and onboard flash, since these simple flash chips don't have smart controllers like ssds, which permutate the sectors. 

In other words: I'm not really that happy with this automatted behaviour without noticing the one installing it, especially because – as I believe – the regular Ubuntu, which uses a different installer – does not do that, and it's not a standard behaviour. Installing hard disks and SSDs differently is not obvious.  

Sure, it's a simple workaround to just put a # in /etc/fstab (no, not that simple, you also need to chmod the unterlying directory accordingly), but you need to know that before. Usually there's not much reason to carefully read fstab after installation.

**Comment by kkofler** [comment]
I think putting `/tmp` on `tmpfs` and then creating a swap file on the SSD is actually counterproductive if you want to protect the SSD's lifetime, you are probably better off with `/tmp` on the SSD and no swap.

For the record, Fedora has defaulted to always (no matter what disk technology is in use, and also no matter what you put in `fstab` because it's implemented as a systemd unit) put `/tmp` on `tmpfs` for a few years now. When the feature was first introduced, I used to disable it by masking that systemd unit, but now that I have 16 GiB of RAM on my desktop computer (and I have been using this computer for a few years now, the `/tmp` on `tmpfs` feature is even older), `/tmp` on `tmpfs` actually works great. I do have a swap space on a HDD RAID0 (the same 2 HDDs mainly have a RAID1 for data, only 16 GiB of each are reserved for a 32 GiB RAID0 for swap), but it is almost always completely empty. So I could just as well not have swap space, I suppose.

**Comment by kkofler** [comment]
That said, I think that the decision whether to write a `tmpfs` line for `/tmp` to `/etc/fstab` really needs to be configurable by the distro. It does not make sense to write such a line on Fedora because there is already a systemd unit that does that. Other distros may want to disable this feature altogether. It should not be hardcoded in Calamares.

**Comment by adriaandegroot** [comment]
Seems like a perfectly sensible thing to PR. A successful PR would:

- [ ] Introduce a key *tmpOptions* into `fstab.conf`; this could have sub-keys, e.g. *default* and *ssd*, which then have at least a boolean *tmpfs* (set to true or false)
- [ ] Introduce documentation for that key, in `fstab.conf`
- [ ] Add the key into the schema file
- [ ] Change the logic in `main.py` to check whether the disk is SSD, then lookup the relevant *tmpOptions* subkey, and apply relevant options

Default when unset / broken should be "tmpfs for SSD, nothing for others" (to preserve current behavior).

**Comment by abalfoort** [comment]
You mean like this in fstab.conf:
```
tmpOptions:
    default: tmpfs=false
    ssd: tmpfs=true;options=defaults,noatime,mode=1777
```
 Then split on ; and = to get the values?
 Is this configurable for ssd only, or if the distro maintainer so chooses, even for hd (by defining default: tmpfs=true)?

**Comment by dalto8** [comment]
> Then split on ; and = to get the values?

The conf files are in yaml.  Calamares parses the yaml for you.

Take a look at [netinstall.conf](https://github.com/calamares/calamares/blob/calamares/src/modules/netinstall/netinstall.conf) for an example with lots of subkeys.

As that config is now very different in 3.3.  I wonder if it would be better to introduce it there?  In 3.3 most of the filesystem mounting configuration has been moved to mount.conf.

**Comment by adriaandegroot** [comment]
More typical would be (I'm following dalto8's comments, addressed to abalfoort):

```
tmpOptions:
    default:
        tmpfs: false
        options: ""
    ssd:
        tmpfs: true
        options: "defaults,noatime,mode=1777"
```

Then YAML (and the schema) does some of the lifting for you, and code would look at more-specific keys in the configuration.

**Comment by abalfoort** [comment]
Thanks, I'll look into it after I finished some other partitioning bugs I found.
